pkg/controller/pravegacluster/upgrade.go

```diff
@@ -108,9 +115,60 @@ func (r *ReconcilePravegaCluster) clearUpgradeStatus(p *pravegav1alpha1.PravegaC
 	return nil
 }
 
-func (r *ReconcilePravegaCluster) syncComponentsVersion(p *pravegav1alpha1.PravegaCluster) (err error) {
-	var synced bool
+func (r *ReconcilePravegaCluster) rollbackClusterVersion(p *pravegav1alpha1.PravegaCluster, version string) (err error) {
+	_, rollbackCondition := p.Status.GetClusterCondition(pravegav1alpha1.ClusterConditionRollback)
+	if rollbackCondition == nil {
+		// We're in the first iteration for Rollback
+		// Add Rollback Condition to Cluster Status
+		p.Status.SetRollbackConditionTrue()
+		p.Spec.Version = version
+		p.Status.TargetVersion = p.Spec.Version
+		updateErr := r.client.Status().Update(context.TODO(), p)
+		if updateErr != nil {
+			log.Printf("Error updating cluster: %v", updateErr.Error())
+			return fmt.Errorf("Error updating cluster status. %v", updateErr)
+		}
+	}
+
+	syncCompleted, err := r.syncComponentsVersion(p)
```

Comment:
For the rollback, I'm not sure the execution sequence should be the same as the upgrade. Say we have an upgraded bookkeeper client that has some new features in the updated segmentstore. If we roll back the bookkeeper server first, is it possible that the segmentstore has errors due to its bookkeeper client code uses the new feature against an old version bookkeeper server? It occurs to me because in the bookkeeper [doc](https://bookkeeper.apache.org/docs/4.7.2/admin/upgrade/), it says usually we need to upgrade server before client. So for rollback, the sequence probably should be reversed somehow?

Reply:
Yes. That sound like a valid point. Will consider reversing the order

Reply:
Changed Rollback to updates components in reverse order as upgrade.